ws.c

```diff
@@ -0,0 +1,409 @@
+	Base64encode(out, digest, sizeof(digest));
+	return out;
+}
+
+static void
+close_after_write_cb(struct bufferevent *bev, void *ctx)
+{
+	if (evbuffer_get_length(bufferevent_get_output(bev)) == 0) {
+		evws_connection_free(ctx);
+	}
+}
+
+static void
+close_event_cb(struct bufferevent *bev, short what, void *ctx)
+{
+	evws_connection_free(ctx);
+}
+
+void
+evws_close(struct evws_connection *evws, uint16_t reason)
+{
+	uint8_t fr[4] = {0x8 | 0x80, 2, 0};
+	struct evbuffer *output;
+	uint16_t *u16;
+
+	if (evws->closed)
+		return;
+	evws->closed = true;
+
+	u16 = (uint16_t *)&fr[2];
+	*u16 = htons((int16_t)reason);
+	output = bufferevent_get_output(evws->bufev);
+	evbuffer_add(output, fr, 4);
+
+	/* wait for close frame writing complete and close connection */
+	bufferevent_setcb(
+		evws->bufev, NULL, close_after_write_cb, close_event_cb, evws);
+}
+
+static void
+evws_force_disconnect_(struct evws_connection *evws)
+{
+	evws_close(evws, WS_CR_NONE);
+}
+
+static enum WebSocketFrameType
+get_ws_frame(unsigned char *in_buffer, int buf_len, unsigned char **payload_ptr,
+	int *out_len)
+{
+	unsigned char opcode;
+	unsigned char fin;
+	unsigned char masked;
+	int payload_len;
+	int pos;
+	int length_field;
+	unsigned int mask;
+
+	if (buf_len < 3) {
+		return INCOMPLETE_FRAME;
+	}
+
+	opcode = in_buffer[0] & 0x0F;
+	fin = (in_buffer[0] >> 7) & 0x01;
+	masked = (in_buffer[1] >> 7) & 0x01;
+
+	payload_len = 0;
+	pos = 2;
+	length_field = in_buffer[1] & (~0x80);
+	mask = 0;
+
+	if (length_field <= 125) {
+		payload_len = length_field;
+	} else if (length_field == 126) { /* msglen is 16bit */
+		payload_len = ntohs(*(uint16_t *)(in_buffer + 2));
+		pos += 2;
+	} else if (length_field == 127) { /* msglen is 64bit */
+		payload_len = ntohs(*(uint64_t *)(in_buffer + 2));
+		pos += 8;
+	}
+	if (buf_len < payload_len + pos) {
+		return INCOMPLETE_FRAME;
+	}
+
+	if (masked) {
+		unsigned char *c;
+
+		mask = *((unsigned int *)(in_buffer + pos));
+		pos += 4;
+
+		/* unmask data */
+		c = in_buffer + pos;
+		for (int i = 0; i < payload_len; i++) {
+			c[i] = c[i] ^ ((unsigned char *)(&mask))[i % 4];
+		}
+	}
+
+	*payload_ptr = in_buffer + pos;
+	*out_len = payload_len;
+
+	switch (opcode) {
+	case 0x0:
+		return fin ? TEXT_FRAME : INCOMPLETE_TEXT_FRAME;
+	case 0x1:
+		return fin ? TEXT_FRAME : INCOMPLETE_TEXT_FRAME;
+	case 0x2:
+		return fin ? BINARY_FRAME : INCOMPLETE_BINARY_FRAME;
+	case 0x9:
+		return PING_FRAME;
+	case 0xA:
+		return PONG_FRAME;
+	default:
+		return ERROR_FRAME;
+	}
+}
+
+
+static void
+ws_evhttp_read_cb(struct bufferevent *bufev, void *arg)
+{
+	struct evws_connection *evws = arg;
+	unsigned char *payload;
+	enum WebSocketFrameType type;
+	int msg_len, in_len;
+	struct evbuffer *input = bufferevent_get_input(evws->bufev);
+
+	while ((in_len = evbuffer_get_length(input))) {
+		unsigned char *data = evbuffer_pullup(input, in_len);
+
+		type = get_ws_frame(data, in_len, &payload, &msg_len);
+
+		if (type == TEXT_FRAME) {
+			if (evws->cb) {
```

Comment:
BTW is it possible to not have cb? And if so, does it make sense?

Reply:
Didn't get this part, what do you mean by that? Libevent somehow needs to deliver parsed messages to the user, and I guess the callback is the most common way to do this. The same idea is used in WebSocket browser API https://developer.mozilla.org/en-US/docs/Web/API/WebSocket

Reply:
Seems that I badly explained it, I meant the condition itself `evws->cb`, is it possible for it to be `false`? And if so, does this makes sense?

IOW I'm suggesting removing it, but not sure that it is safe now.

Reply:
Right, it's redundant because is set in `evws_new_session` and there is no way to avoid this set, removing